### PR TITLE
OpenTofu Installer

### PR DESCRIPTION
**Pull request by janosdebugs**
This PR implements the OpenTofu Installer for Powershell/POSIX shells on Windows, Linux, MacOS and hopefully *NIX.

The intention is to support all POSIX compliant shells as well as Powershell on Windows 10 and 11 to install OpenTofu via all documented means on the website.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,18 +1,26 @@
...
+Pages. The installation scripts are located in [`src/install-opentofu.sh`](src/install-opentofu.sh) (POSIX) and [`src/install-opentofu.ps1`](src/install-opentofu.ps1) (Powershell). The Cloudflare function managing the MIME type assignment is located in
 [`src/functions/index.ts`](src/functions/index.ts).
 
-## Testing the script (Linux only, WIP)
+## Testing the script
 
-You can test the [installation script](src/install.sh) manually, or you can use `docker compose` to run the automated
+### Linux
+
+You can test the installation scriptmanually, or you can use `docker compose` to run the automated
```
Comment (Yantrio) [inline README.md:11]:
```suggestion
You can test the installation script manually, or you can use `docker compose` to run the automated

```

File: src/install-opentofu.sh [hunk 2]
```diff
@@ -0,0 +1,1206 @@
...
+    fi
+  fi
+  if command_exists "wget"; then
+    log_debug "Downloading using wget..."
+    if [ "${IS_GITHUB}" -eq 1 ]; then
+      if ! wget -q --header="Authorization: token ${GITHUB_TOKEN}" -O "$2" "$1"; then
+        log_debug "Download failed."
+        return $TOFU_INSTALL_RETURN_CODE_DOWNLOAD_FAILED
+      fi
+    else
```
Comment (Yantrio) [inline src/install-opentofu.sh:243]: Might be nice to have a log in here mentioning that it's using the github token because it was detected.

Otherwise some people may get confused if it fails

File: src/install-opentofu.sh [hunk 3]
```diff
@@ -0,0 +1,1206 @@
...
+  return $TOFU_INSTALL_EXIT_CODE_OK
+}
+
+# This function installs OpenTofu via Homebrew. It returns $TOFU_INSTALL_EXIT_CODE_INSTALL_METHOD_NOT_SUPPORTED if
+# Homebrew is not available.
+install_brew() {
+  if ! command_exists "brew"; then
+    return $TOFU_INSTALL_EXIT_CODE_INSTALL_METHOD_NOT_SUPPORTED
+  fi
+  log_info "Installing OpenTofu using Homebrew..."
```
Comment (Yantrio) [inline src/install-opentofu.sh:673]: maybe we should run `brew update` here first?

**Review by Yantrio — APPROVED** [review]
I've reviewed the bash and tests, but not the ps1. I have very little knowledge of powershell and wouldn't feel comfortable signing off on that. Sorry!


Great work overall though!

**Review by Yantrio — CHANGES_REQUESTED** [review]
I've reviewed the bash and tests, but not the ps1. I have very little knowledge of powershell and wouldn't feel comfortable signing off on that. Sorry!


Great work overall though!

**Review by Yantrio — APPROVED** [review]

**Review by cam72cam — APPROVED** [review]

File: .github/workflows/test.yml [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+      - name: Install Cosign
+        uses: sigstore/cosign-installer@v3.3.0
+      - name: Test
+        env:
+          DISTRO: ${{ matrix.distro }}
+          METHOD: ${{ matrix.method }}
+          SH: ${{ matrix.shell }}
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+        run: |
+          cd tests/linux
```
Comment (tom-reinders) [inline .github/workflows/test.yml:28]: Why use a mixture of `cd` and working-directory in windows?

Why not use working-directory everywhere?

Comment (janosdebugs) [reply .github/workflows/test.yml:28]: Good catch, let me fix that!

Comment (janosdebugs) [reply .github/workflows/test.yml]: Implemented, thanks @tom-reinders !

**Review by adamconnelly — COMMENTED** [review]
I've just taken a look at the Powershell install script as per @Yantrio's request.

Looks good overall (although just for the record my Powershell skills are definitely very rusty). I left some comments, but for the most part it's all pretty minor.

File: src/install-opentofu.ps1 [hunk 5]
```diff
@@ -0,0 +1,502 @@
...
+    [Parameter(Mandatory = $false)]
+    [switch]$help = $false,
+    [Parameter(Mandatory = $false)]
+    [string]$installPath = "",
+    [Parameter(Mandatory = $false)]
+    [string]$opentofuVersion = "latest",
+    [Parameter(Mandatory = $false)]
+    [string]$installMethod,
+    [Parameter(Mandatory = $false)]
+    [string]$cosignPath = "cosign.exe",
```
[the remark this answers is not in the document]
Comment (janosdebugs) [reply src/install-opentofu.ps1:49]: I've done this so the defaults appear in the `Get-Help` output. Unfortunately, some defaults are dynamic (such as the `installPath` default depending on the presence of the `allUsers` flag) and there is no good way to do this. The default values are also needed for the usage documentation.

File: src/install-opentofu.ps1 [hunk 6]
```diff
@@ -0,0 +1,502 @@
...
+    Write-Warning "${orange}${message}${normal}"
+}
+
+function logError() {
+    param(
+        $message
+    )
+    try
+    {
+        [Console]::Error.WriteLine("${red}${message}${normal}")
```
[the remark this answers is not in the document]
Comment (janosdebugs) [reply src/install-opentofu.ps1:161]: It seemed to me that `Write-Error` actually only creates an exception and doesn't write to the standard error. It also makes a nasty-looking error that's hard to parse, whereas we really only need that one sentence. Maybe there's a better way to do this, but I don't know.

File: src/install-opentofu.ps1 [hunk 7]
```diff
@@ -0,0 +1,502 @@
...
+
+    logInfo "Performing standalone installation to ${installPath}..."
+
+    if (!$skipVerify) {
+        logInfo("Checking if cosign is available...")
+        $cosignError = "Cosign is not installed but required for the standalone installation. Please install cosign / provide the cosign path with the -cosignPath parameter, disable integrity verification with -skipVerify (not recommended), or select a different installation method."
+        try {
+            $ErrorActionPreference = 'stop'
+            if(!(Get-Command $cosignPath)){
+                throw [InstallMethodNotSupportedException]::new($cosignError)
```
[the remark this answers is not in the document]
Comment (janosdebugs) [reply src/install-opentofu.ps1:238]: Hmm you're right. Maybe `InstallRequirementNotMetException`?

File: src/install-opentofu.ps1 [hunk 8]
```diff
@@ -0,0 +1,502 @@
...
+        if (!$skipVerify)
+        {
+            if ($cosignIdentity -eq "autodetect") {
+                if ($opentofuVersion -in "1.6.0-beta4","1.6.0-beta3","1.6.0-beta2","1.6.0-beta1","1.6.0-alpha5","1.6.0-alpha4","1.6.0-alpha3","1.6.0-alpha2","1.6.0-alpha1") {
+                    $cosignIdentity = "https://github.com/opentofu/opentofu/.github/workflows/release.yml@refs/tags/v${OPENTOFU_VERSION}"
+                } else {
+                    if ($opentofuVersion.Contains("alpha") -or $opentofuVersion.Contains("beta")) {
+                        $cosignIdentity="https://github.com/opentofu/opentofu/.github/workflows/release.yml@refs/heads/main"
+                    } else {
+                        $ver = [version]($opentofuVersion -replace "-alpha|-beta|-rc.*")
```
[the remark this answers is not in the document]
Comment (janosdebugs) [reply src/install-opentofu.ps1:331]: Whoops, you are correct!

File: src/install-opentofu.ps1 [hunk 9]
```diff
@@ -0,0 +1,502 @@
...
+            # Also note that when using Start-Process with RunAs, it is not possible to pass environment variables.
+            # (The documentation is lying!)
+            #
+            # TL;DR: Make sure to MANUALLY test privilege elevation if you change this as we can't test UAC in CI! Make
+            # sure to test with paths containing spaces too! Did I mention to test this? Test it. I'm serious.
+
+            logInfo "Unpacking with elevated privileges..."
+            $logDir = tempdir
+            # TODO capture the log output of the shell running as admin and clean up afterwards.
+            $argList = @("-NonInteractive", "-File", ($scriptCommand | escapePathArgument), "-internalContinue", "-allUsers", "-installMethod", "standalone", "-installPath", ($installPath | escapePathArgument), "-internalZipFile", ($internalZipFile | escapePathArgument))
```
[the remark this answers is not in the document]
Comment (janosdebugs) [reply src/install-opentofu.ps1:366]: Since this is deep inside the standalone installer this is the only method that makes sense. When we add more installation methods we will need to refactor this and provide a generic UAC privilege elevation method.

**Comment by janosdebugs** [comment]
Thanks for your help @adamconnelly !

File: src/install-opentofu.ps1 (hunk 6 above)
[the remark this answers is not in the document]
Comment (adamconnelly) [reply src/install-opentofu.ps1:161]: No worries - makes sense.

File: src/install-opentofu.ps1 (hunk 7 above)
[the remark this answers is not in the document]
Comment (adamconnelly) [reply src/install-opentofu.ps1:238]: Sure - that works.

File: src/install-opentofu.ps1 (hunk 5 above)
[the remark this answers is not in the document]
Comment (adamconnelly) [reply src/install-opentofu.ps1:49]: Yeah - setting the defaults here so that they appear in the help output makes sense. I just mean that maybe instead of doing stuff like this later in the script:

```powershell
if (!$cosignPath) {
    $cosignPath = $defaultCosignPath
}
```

You could actually just fail with an error message if no path is specified. That way the default is only set in a single place.

I totally get that you can't do it for stuff like `installPath`.

[the remark this answers is not in the document]
Comment (janosdebugs) [reply src/install-opentofu.ps1:49]: Cosign doesn't have any information on how to install it on Windows, which is a shame. I have it installed in the path, which is why I built it this way. Not sure how others use it, maybe that's something we can improve based on user feedback? We check if this path is valid anyway, if it isn't the user will get an error message indicating that.

File: src/install-opentofu.ps1 [hunk 10]
```diff
@@ -0,0 +1,502 @@
...
+        if ($allUsers) {
+            logInfo "Updating system PATH variable..."
+            $target = [EnvironmentVariableTarget]::Machine
+        } else {
+            logInfo "Updating user PATH variable..."
+            $target = [EnvironmentVariableTarget]::User
+        }
+        $currentPath = [Environment]::GetEnvironmentVariable("Path", $target)
+        if (!($currentPath.Contains($installPath))) {
+            [Environment]::SetEnvironmentVariable("Path", $currentPath + ";${installPath}", $target)
```
Comment (peterdeme) [inline src/install-opentofu.ps1:213]: question: are you sure this permanently changes the env var and not just temporarily for the current session? 🤔

Comment (janosdebugs) [reply src/install-opentofu.ps1:213]: Hi @peterdeme , I have tested this method and it does indeed add the path permanently to the user's or the system profile depending on the `$target` on Windows 11. The SO question is where I found the hint for this API call, I just adapted the call to our needs.

Screenshot of the Windows path editor window

**Review by peterdeme — APPROVED** [review]
